auth/store.go

```diff
@@ -1070,3 +1073,40 @@ func NewTokenProvider(tokenOpts string, indexWaiter func(uint64) <-chan struct{}
 		return nil, ErrInvalidAuthOpts
 	}
 }
+
+func (as *authStore) WithRoot(ctx context.Context) context.Context {
+	if !as.isAuthEnabled() {
+		return ctx
+	}
+
+	var ctxForAssign context.Context
+	if ts, ok := as.tokenProvider.(*tokenSimple); ok && ts != nil {
+		ctx1 := context.WithValue(ctx, "index", uint64(0))
+		prefix, err := ts.genTokenPrefix()
+		if err != nil {
+			plog.Errorf("failed to generate prefix of internally used token")
+			return ctx
+		}
+		ctxForAssign = context.WithValue(ctx1, "simpleToken", prefix)
+	} else {
+		ctxForAssign = ctx
+	}
+
+	token, err := as.tokenProvider.assign(ctxForAssign, "root", as.Revision())
+	if err != nil {
+		// this must not happen
+		plog.Errorf("failed to assign token for lease revoking: %s", err)
+		return ctx
+	}
+
+	mdMap := map[string]string{
+		"token": token,
+	}
+	tokenMD := metadata.New(mdMap)
+
+	// clean up tls info to ensure using root credential
+	ctx = peer.NewContext(ctx, nil)
```

Comment:
I clean up the tls info in the context to ensure using the assigned token rather than CN based auth. But I'm not quite sure if this will cause other problems (at least it's OK during my test), do you have a better suggestion? Thanks. @mitake

Reply:
I see. Let me take a look later (sorry, I'll be busy for a few days). The problem needs serious design consideration.

Reply:
OK, BTW. I just noticed that the priorities of CN based auth and token based auth were exchanged in etcd3.3, so we needn't clean up the tls info if we port you this commit 1296281b27c8fec9a89ba02e104298dec9e4464a to etcd3.2, maybe it's also a solution,  the disadvantage is that this modification will affect all APIs that require authentication. Considering the principle of minimal change, I think that just for fixing this upgrade bug, cleaning up tls info in `WithRoot` may be a better choice.

Reply:
Got it, but let me think about backporting the commit to 3.2, because it might affect users.